src/de.rs

```diff
@@ -262,8 +266,13 @@ impl<'de, 'a, 'b> serde::Deserializer<'de> for &'a mut Deserializer<'b> {
     where
         V: Visitor<'de>,
     {
-        let marker = self.read_marker()?;
-        self.dispatch_based_on_marker(marker, visitor)
+        if !self.read_header {
+            self.read_header = true;
+            visitor.visit_map(MapAccess::with_varint_encoded_fields(self)?)
+        } else {
+            let marker = self.read_marker()?;
+            self.dispatch_based_on_marker(marker, visitor)
+        }
```

Comment:
Nit: Can we use early-return please.

```suggestion
            return visitor.visit_map(MapAccess::with_varint_encoded_fields(self)?)
        }
        
        let marker = self.read_marker()?;
        self.dispatch_based_on_marker(marker, visitor)
```